share/github-backup-utils/ghe-backup-repositories

```diff
@@ -362,4 +364,15 @@ RULES
 done
 bm_end "$(basename $0) - Special Data Directories Sync"
 
+if [ -z "$GHE_SKIP_ROUTE_VERIFICATION" ]; then
+  bm_start "$(basename $0) - Verifying Routes"
+
+  cat $tempdir/*.rsync | sort | uniq > $tempdir/source_routes
+  (cd $backup_dir/ && find * -mindepth 5 -maxdepth 6 -type d -name \*.git -exec dirname {} \; | sort | uniq) > $tempdir/destination_routes
+
+  git diff -- $tempdir/source_routes $tempdir/destination_routes || echo "Warning: One or more repository networks and/or gists were not found on the source appliance. Please contact GitHub Enterprise Support for assistance."
```

Comment:
@snh I suspect these warnings go usually unnoticed. Would it make sense to consider this a `failure` and process it accordingly? https://github.com/github/backup-utils/blob/0b42e114d2e7d47f81f565ae1d625f3d9098fdae/bin/ghe-backup#L229-L236

Reply:
TBH I don't understand that completely yet. Doesn't this check only ensure that all repo **directories** are present on the appliance and on the backup? If `rsync` fails with a 23 error and a single file is not transferred, then we wouldn't learn about it, right?

I saw a number of those errors in the wild:
```
Jun 04 21:39:30 rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1505) [generator=3.0.6]
```

Reply:
Plus, wouldn't that check always generate a false warning if repositories have been deleted during the backup run? Do we delete repositories while the backup is running?

Reply:
> Would it make sense to consider this a failure and process it accordingly?

These errors indicate a consistency issue on the appliance rather than an issue with the backup itself, therefore the backup is still to be considered valid, and not a failure.

> TBH I don't understand that completely yet. Doesn't this check only ensure that all repo directories are present on the appliance and on the backup?

It is intended to check that all routes exist in the backup, it is not intended to detect individual file transfer errors from `rsync`, which should instead be detected through other means.

> Do we delete repositories while the backup is running?

We pause purging during the relevant stage of the backup. This is not an issue.